pymc3/model.py

```diff
@@ -912,51 +913,90 @@ def independent_vars(self):
+            DeprecationWarning,
+        )
+        return self.get_initial_point(recompute=False)
 
-    @property
-    def initial_values(self) -> Dict[TensorVariable, np.ndarray]:
-        """Maps transformed variables to initial values.
+    def get_initial_point(self, recompute: bool = False) -> Dict[str, np.ndarray]:
+        """Gets numeric initial values for all free model variables.
 
-        ⚠ The keys are NOT the objects returned by, `pm.Normal(...)`.
-        For a name-based dictionary use the `initial_point` property.
+        Parameters
+        ----------
+        recompute : bool
+            If `False` cached initial values are returned unless the set of free variables changed.
+            Note that re-using cached numeric initvals can still lead to (shape) problems
+            if shared variables (e.g. `pm.Data`) changed in value or length!
+
+        Returns
+        -------
+        initial_point : dict
+            Maps transformed free variable names to transformed, numeric initial values.
         """
-        return self._initial_values
+        if not recompute and set(self._cached_numeric_initvals) == set(self.initial_values):
+            # Recomputation is off and the set of free variables is unchanged.
+            # Re-using these cached numeric initvals can still lead to shape problems
+            # if dimensions in the model were resized!
+            numeric_initvals = self._cached_numeric_initvals
+        else:
+            numeric_initvals = {}
+            # The entries in `initial_values` are already in topological order and can be evaluated one by one.
+            for rv_value, initval in self.initial_values.items():
+                rv_var = self.values_to_rvs[rv_value]
+                transform = getattr(rv_value.tag, "transform", None)
+                if isinstance(initval, np.ndarray) and transform is None:
+                    # Only untransformed, numeric initvals can be taken as they are.
+                    numeric_initvals[rv_value] = initval
+                else:
+                    # Evaluate initvals that are None, symbolic or need to be transformed.
+                    # They can depend on other initvals from higher up in the graph,
+                    # which are therefore fed to the evaluation as "givens".
+                    test_value = getattr(rv_var.tag, "test_value", None)
+                    numeric_initvals[rv_value] = self._eval_initval(
+                        rv_var, initval, test_value, transform, given=numeric_initvals
+                    )
 
-    @property
-    def disc_vars(self):
-        """All the discrete variables in the model"""
-        return list(typefilter(self.value_vars, discrete_types))
+            # Cache the evaluation results for next time.
+            self._cached_numeric_initvals = numeric_initvals
+        return Point(list(numeric_initvals.items()), model=self)
```

Comment:
Were we creating a `Point` object every time we accessed `initial_point`before?

Reply:
Yeah.

I considered to cache the point, but the set comparison further up is a little easier with the tensors as keys instead of the names.

Reply:
Caching was improved in #4982